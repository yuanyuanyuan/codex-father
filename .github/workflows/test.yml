name: Test Suite

on:
  push:
    branches: [ main, develop, '001-2-specification-data' ]
  pull_request:
    branches: [ main, develop ]
  schedule:
    # 每天凌晨2点运行完整测试
    - cron: '0 2 * * *'

jobs:
  test-unit:
    name: Unit Tests
    runs-on: ubuntu-latest
    
    strategy:
      matrix:
        node-version: [18.x, 20.x]
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Run type check
      run: npm run typecheck
      
    - name: Run linter
      run: npm run lint:check
      
    - name: Run unit tests
      run: npm run test:unit
      
    - name: Run unit tests with coverage
      run: npm run test:coverage:unit
      
    - name: Upload coverage to Codecov
      uses: codecov/codecov-action@v3
      with:
        file: ./coverage/lcov.info
        flags: unit
        name: codecov-unit

  test-integration:
    name: Integration Tests
    runs-on: ubuntu-latest
    needs: test-unit
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js
      uses: actions/setup-node@v4
      with:
        node-version: '20.x'
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Build project
      run: npm run build
      
    - name: Run integration tests
      run: npm run test:integration
      
    - name: Run contract tests
      run: npm run test:contract

  test-e2e:
    name: E2E Tests
    runs-on: ubuntu-latest
    needs: [test-unit, test-integration]
    
    strategy:
      matrix:
        node-version: [18.x, 20.x]
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Build project
      run: npm run build
      
    - name: Check E2E environment
      run: npm run test:e2e:check
      
    - name: Prepare E2E environment
      run: npm run test:e2e:prepare
      
    - name: Run E2E tests
      run: npm run test:e2e
      
    - name: Run E2E tests with coverage
      run: npm run test:e2e:coverage
      
    - name: Upload E2E coverage
      uses: codecov/codecov-action@v3
      with:
        file: ./coverage/e2e/lcov.info
        flags: e2e
        name: codecov-e2e
      
    - name: Generate test report
      run: npm run test:report
      
    - name: Upload test artifacts
      uses: actions/upload-artifact@v3
      if: always()
      with:
        name: e2e-test-results-${{ matrix.node-version }}
        path: |
          test-results/
          coverage/
        retention-days: 7

  test-performance:
    name: Performance Tests
    runs-on: ubuntu-latest
    needs: test-e2e
    if: github.event_name == 'push' || github.event_name == 'schedule'
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js
      uses: actions/setup-node@v4
      with:
        node-version: '20.x'
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Build project
      run: npm run build
      
    - name: Run performance benchmarks
      run: npm run test:performance
      
    - name: Upload performance results
      uses: actions/upload-artifact@v3
      with:
        name: performance-results
        path: test-results/performance-*.json

  security-scan:
    name: Security Scan
    runs-on: ubuntu-latest
    needs: test-unit
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js
      uses: actions/setup-node@v4
      with:
        node-version: '20.x'
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Run security audit
      run: npm audit --audit-level high
      
    - name: Run CodeQL Analysis
      uses: github/codeql-action/analyze@v2
      with:
        languages: javascript

  build-test:
    name: Build Test
    runs-on: ${{ matrix.os }}
    needs: test-unit
    
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        node-version: [18.x, 20.x]
        exclude:
          - os: windows-latest
            node-version: 18.x
    
    steps:
    - name: Checkout code
      uses: actions/checkout@v4
      
    - name: Setup Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        
    - name: Install dependencies
      run: npm ci
      
    - name: Build project
      run: npm run build
      
    - name: Test CLI installation
      run: npm pack && npm install -g codex-father-*.tgz
      
    - name: Test basic CLI commands
      run: |
        codex-father --version
        codex-father --help

  test-summary:
    name: Test Summary
    runs-on: ubuntu-latest
    needs: [test-unit, test-integration, test-e2e, test-performance]
    if: always()
    
    steps:
    - name: Download all artifacts
      uses: actions/download-artifact@v3
      
    - name: Generate summary report
      run: |
        echo "# Test Summary" >> $GITHUB_STEP_SUMMARY
        echo "" >> $GITHUB_STEP_SUMMARY
        echo "## Test Results" >> $GITHUB_STEP_SUMMARY
        echo "- Unit Tests: ${{ needs.test-unit.result }}" >> $GITHUB_STEP_SUMMARY
        echo "- Integration Tests: ${{ needs.test-integration.result }}" >> $GITHUB_STEP_SUMMARY
        echo "- E2E Tests: ${{ needs.test-e2e.result }}" >> $GITHUB_STEP_SUMMARY
        echo "- Performance Tests: ${{ needs.test-performance.result }}" >> $GITHUB_STEP_SUMMARY
        echo "" >> $GITHUB_STEP_SUMMARY
        echo "## Coverage Reports" >> $GITHUB_STEP_SUMMARY
        echo "Coverage reports have been uploaded to Codecov." >> $GITHUB_STEP_SUMMARY
        
    - name: Overall status
      if: contains(needs.*.result, 'failure')
      run: |
        echo "❌ Some tests failed. Check the individual job results for details."
        exit 1
      else
      run: |
        echo "✅ All tests passed successfully!"

  notify:
    name: Notify Results
    runs-on: ubuntu-latest
    needs: [test-summary]
    if: failure() && (github.event_name == 'push' || github.event_name == 'pull_request')
    
    steps:
    - name: Notify on failure
      uses: actions/github-script@v6
      with:
        script: |
          const { owner, repo } = context.repo;
          const sha = context.sha;
          
          await github.rest.repos.createCommitStatus({
            owner,
            repo,
            sha,
            state: 'failure',
            target_url: 'https://github.com/${owner}/${repo}/actions/runs/${context.runId}',
            description: 'Tests failed',
            context: 'ci/test'
          });