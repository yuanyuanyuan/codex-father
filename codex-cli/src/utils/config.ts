// NOTE: We intentionally point the TypeScript import at the source file
// (`./auto-approval-mode.ts`) instead of the emitted `.js` bundle.  This makes
// the module resolvable when the project is executed via `ts-node`, which
// resolves *source* paths rather than built artefacts.  During a production
// build the TypeScript compiler will automatically rewrite the path to
// `./auto-approval-mode.js`, so the change is completely transparent for the
// compiled `dist/` output used by the published CLI.

import type { FullAutoErrorMode } from "./auto-approval-mode.js";

import { log, isLoggingEnabled } from "./agent/log.js";
import { AutoApprovalMode } from "./auto-approval-mode.js";
import { reportMissingAPIKeyForProvider } from "./model-utils.js";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { load as loadYaml, dump as dumpYaml } from "js-yaml";
import { homedir } from "os";
import { dirname, join, extname, resolve as resolvePath } from "path";

export const DEFAULT_APPROVAL_MODE = AutoApprovalMode.SUGGEST;
export const DEFAULT_INSTRUCTIONS = "";

export const CONFIG_DIR = join(homedir(), ".codex");
export const CONFIG_JSON_FILEPATH = join(CONFIG_DIR, "config.json");
export const CONFIG_YAML_FILEPATH = join(CONFIG_DIR, "config.yaml");
export const CONFIG_YML_FILEPATH = join(CONFIG_DIR, "config.yml");

// Keep the original constant name for backward compatibility, but point it at
// the default JSON path. Code that relies on this constant will continue to
// work unchanged.
export const CONFIG_FILEPATH = CONFIG_JSON_FILEPATH;
export const INSTRUCTIONS_FILEPATH = join(CONFIG_DIR, "instructions.md");

export const OPENAI_TIMEOUT_MS =
  parseInt(process.env["OPENAI_TIMEOUT_MS"] || "0", 10) || undefined;

export let DEFAULT_PROVIDER = "openai";
export let API_KEY = "";

// Gracefully fallback to a provider if we have a missing API key.
if (!process.env["OPENAI_API_KEY"]) {
  if (process.env["GOOGLE_GENERATIVE_AI_API_KEY"]) {
    DEFAULT_PROVIDER = "gemini";
  } else if (process.env["OPENROUTER_API_KEY"]) {
    DEFAULT_PROVIDER = "openrouter";
  } else if (process.env["XAI_API_KEY"]) {
    DEFAULT_PROVIDER = "xai";
  }
}

function getAPIKeyForProviderOrExit(provider: string): string {
  switch (provider) {
    case "openai":
      if (process.env["OPENAI_API_KEY"]) {
        return process.env["OPENAI_API_KEY"];
      }
      reportMissingAPIKeyForProvider(provider);
      process.exit(1);
      break;
    case "gemini":
      if (process.env["GOOGLE_GENERATIVE_AI_API_KEY"]) {
        return process.env["GOOGLE_GENERATIVE_AI_API_KEY"];
      }
      reportMissingAPIKeyForProvider(provider);
      process.exit(1);
      break;
    case "openrouter":
      if (process.env["OPENROUTER_API_KEY"]) {
        return process.env["OPENROUTER_API_KEY"];
      }
      reportMissingAPIKeyForProvider(provider);
      process.exit(1);
      break;
    case "ollama":
      // Ollama doesn't require an API key but the openai client requires one
      return "ollama";
    case "xai":
      if (process.env["XAI_API_KEY"]) {
        return process.env["XAI_API_KEY"];
      }
      reportMissingAPIKeyForProvider(provider);
      process.exit(1);
      break;
    default:
      reportMissingAPIKeyForProvider("");
      process.exit(1);
  }
}

function baseURLForProvider(provider: string): string {
  switch (provider) {
    case "openai":
      return "https://api.openai.com/v1";
    case "ollama":
      return process.env["OLLAMA_BASE_URL"] ?? "http://localhost:11434/v1";
    case "gemini":
      return "https://generativelanguage.googleapis.com/v1beta/openai/";
    case "openrouter":
      return "https://openrouter.ai/api/v1";
    case "xai":
      return "https://api.x.ai/v1";
    default:
      // TODO throw?
      return "";
  }
}

function defaultModelsForProvider(provider: string): {
  agentic: string;
  fullContext: string;
} {
  switch (provider) {
    case "openai":
      return {
        agentic: "o4-mini",
        fullContext: "o3",
      };
    case "gemini":
      return {
        agentic: "gemini-2.5-pro-preview-03-25",
        fullContext: "gemini-2.0-flash",
      };
    case "openrouter":
      return {
        agentic: "openai/o4-mini",
        fullContext: "openai/o3",
      };
    case "xai":
      return {
        agentic: "grok-3-mini-beta",
        fullContext: "grok-3-beta",
      };
    default:
      return {
        agentic: "",
        fullContext: "",
      };
  }
}

export function setApiKey(apiKey: string): void {
  API_KEY = apiKey;
}

// Formatting (quiet mode-only).
export const PRETTY_PRINT = Boolean(process.env["PRETTY_PRINT"] || "");

// Represents config as persisted in config.json.
export type StoredConfig = {
  model?: string;
  baseURL?: string;
  provider?: string;
  approvalMode?: AutoApprovalMode;
  fullAutoErrorMode?: FullAutoErrorMode;
  memory?: MemoryConfig;
};

// Minimal config written on first run.  An *empty* model string ensures that
// we always fall back to DEFAULT_MODEL on load, so updates to the default keep
// propagating to existing users until they explicitly set a model.
export const EMPTY_STORED_CONFIG: StoredConfig = { model: "" };

// Pre‑stringified JSON variant so we don’t stringify repeatedly.
const EMPTY_CONFIG_JSON = JSON.stringify(EMPTY_STORED_CONFIG, null, 2) + "\n";

export type MemoryConfig = {
  enabled: boolean;
};

// Represents full runtime config, including loaded instructions.
export type AppConfig = {
  apiKey?: string;
  baseURL?: string;
  provider?: string;
  model: string;
  instructions: string;
  approvalMode?: AutoApprovalMode;
  fullAutoErrorMode?: FullAutoErrorMode;
  memory?: MemoryConfig;
};

// ---------------------------------------------------------------------------
// Project doc support (codex.md)
// ---------------------------------------------------------------------------

export const PROJECT_DOC_MAX_BYTES = 32 * 1024; // 32 kB

const PROJECT_DOC_FILENAMES = ["codex.md", ".codex.md", "CODEX.md"];

export function discoverProjectDocPath(startDir: string): string | null {
  const cwd = resolvePath(startDir);

  // 1) Look in the explicit CWD first:
  for (const name of PROJECT_DOC_FILENAMES) {
    const direct = join(cwd, name);
    if (existsSync(direct)) {
      return direct;
    }
  }

  // 2) Fallback: walk up to the Git root and look there.
  let dir = cwd;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const gitPath = join(dir, ".git");
    if (existsSync(gitPath)) {
      // Once we hit the Git root, search its top‑level for the doc
      for (const name of PROJECT_DOC_FILENAMES) {
        const candidate = join(dir, name);
        if (existsSync(candidate)) {
          return candidate;
        }
      }
      // If Git root but no doc, stop looking.
      return null;
    }

    const parent = dirname(dir);
    if (parent === dir) {
      // Reached filesystem root without finding Git.
      return null;
    }
    dir = parent;
  }
}

/**
 * Load the project documentation markdown (codex.md) if present. If the file
 * exceeds {@link PROJECT_DOC_MAX_BYTES} it will be truncated and a warning is
 * logged.
 *
 * @param cwd The current working directory of the caller
 * @param explicitPath If provided, skips discovery and loads the given path
 */
export function loadProjectDoc(cwd: string, explicitPath?: string): string {
  let filepath: string | null = null;

  if (explicitPath) {
    filepath = resolvePath(cwd, explicitPath);
    if (!existsSync(filepath)) {
      // eslint-disable-next-line no-console
      console.warn(`codex: project doc not found at ${filepath}`);
      filepath = null;
    }
  } else {
    filepath = discoverProjectDocPath(cwd);
  }

  if (!filepath) {
    return "";
  }

  try {
    const buf = readFileSync(filepath);
    if (buf.byteLength > PROJECT_DOC_MAX_BYTES) {
      // eslint-disable-next-line no-console
      console.warn(
        `codex: project doc '${filepath}' exceeds ${PROJECT_DOC_MAX_BYTES} bytes – truncating.`,
      );
    }
    return buf.slice(0, PROJECT_DOC_MAX_BYTES).toString("utf-8");
  } catch {
    return "";
  }
}

export type LoadConfigOptions = {
  /** Working directory used for project doc discovery */
  cwd?: string;
  /** Disable inclusion of the project doc */
  disableProjectDoc?: boolean;
  /** Explicit path to project doc (overrides discovery) */
  projectDocPath?: string;
  /** Whether we are in fullcontext mode. */
  isFullContext?: boolean;
  /** The provider to use. */
  provider?: string;
  /** Force the API key for testing purposes. */
  forceApiKeyForTest?: string;
};

export const loadInstructions = (
  instructionsPath: string | undefined = INSTRUCTIONS_FILEPATH,
  options: LoadConfigOptions = {},
): string => {
  const instructionsFilePathResolved =
    instructionsPath ?? INSTRUCTIONS_FILEPATH;
  const userInstructions = existsSync(instructionsFilePathResolved)
    ? readFileSync(instructionsFilePathResolved, "utf-8")
    : DEFAULT_INSTRUCTIONS;

  // Project doc support.
  const shouldLoadProjectDoc =
    !options.disableProjectDoc &&
    process.env["CODEX_DISABLE_PROJECT_DOC"] !== "1";

  let projectDoc = "";
  let projectDocPath: string | null = null;
  if (shouldLoadProjectDoc) {
    const cwd = options.cwd ?? process.cwd();
    projectDoc = loadProjectDoc(cwd, options.projectDocPath);
    projectDocPath = options.projectDocPath
      ? resolvePath(cwd, options.projectDocPath)
      : discoverProjectDocPath(cwd);
    if (projectDocPath) {
      if (isLoggingEnabled()) {
        log(
          `[codex] Loaded project doc from ${projectDocPath} (${projectDoc.length} bytes)`,
        );
      }
    } else {
      if (isLoggingEnabled()) {
        log(`[codex] No project doc found in ${cwd}`);
      }
    }
  }

  const combinedInstructions = [userInstructions, projectDoc]
    .filter((s) => s && s.trim() !== "")
    .join("\n\n--- project-doc ---\n\n");

  try {
    // Always ensure the instructions file exists so users can edit it.
    if (!existsSync(instructionsFilePathResolved)) {
      const instrDir = dirname(instructionsFilePathResolved);
      if (!existsSync(instrDir)) {
        mkdirSync(instrDir, { recursive: true });
      }
      writeFileSync(instructionsFilePathResolved, userInstructions, "utf-8");
    }
  } catch {
    // Silently ignore any errors – failure to persist the defaults shouldn't
    // block the CLI from starting.  A future explicit `codex config` command
    // or `saveConfig()` call can handle (re‑)writing later.
  }

  return combinedInstructions;
};

export const loadConfig = (
  configPath: string | undefined = CONFIG_FILEPATH,
  instructionsPath: string | undefined = INSTRUCTIONS_FILEPATH,
  options: LoadConfigOptions = {},
): AppConfig => {
  // Determine the actual path to load. If the provided path doesn't exist and
  // the caller passed the default JSON path, automatically fall back to YAML
  // variants.
  let actualConfigPath = configPath;
  if (!existsSync(actualConfigPath)) {
    if (configPath === CONFIG_FILEPATH) {
      if (existsSync(CONFIG_YAML_FILEPATH)) {
        actualConfigPath = CONFIG_YAML_FILEPATH;
      } else if (existsSync(CONFIG_YML_FILEPATH)) {
        actualConfigPath = CONFIG_YML_FILEPATH;
      }
    }
  }

  let storedConfig: StoredConfig = {};
  if (existsSync(actualConfigPath)) {
    const raw = readFileSync(actualConfigPath, "utf-8");
    const ext = extname(actualConfigPath).toLowerCase();
    try {
      if (ext === ".yaml" || ext === ".yml") {
        storedConfig = loadYaml(raw) as unknown as StoredConfig;
      } else {
        storedConfig = JSON.parse(raw);
      }
    } catch {
      // If parsing fails, fall back to empty config to avoid crashing.
      storedConfig = {};
    }
  }

  const storedProvider =
    storedConfig.provider && storedConfig.provider.trim() !== ""
      ? storedConfig.provider.trim()
      : undefined;

  // Treat empty string ("" or whitespace) as absence so we can fall back to
  // the latest DEFAULT_MODEL.
  const storedModel =
    storedConfig.model && storedConfig.model.trim() !== ""
      ? storedConfig.model.trim()
      : undefined;

  const storedBaseURL =
    storedConfig.baseURL && storedConfig.baseURL.trim() !== ""
      ? storedConfig.baseURL.trim()
      : undefined;

  const providerOrDefault = options.provider ?? DEFAULT_PROVIDER;

  const derivedBaseURL = storedProvider
    ? baseURLForProvider(storedProvider)
    : storedBaseURL ?? baseURLForProvider(providerOrDefault);

  const derivedModels = storedProvider
    ? defaultModelsForProvider(storedProvider)
    : defaultModelsForProvider(providerOrDefault);

  const derivedModel =
    storedModel ||
    (options.isFullContext
      ? derivedModels?.fullContext
      : derivedModels?.agentic);

  const derivedProvider = storedProvider ?? providerOrDefault;
  const apiKeyForProvider =
    options.forceApiKeyForTest ?? getAPIKeyForProviderOrExit(derivedProvider);

  const config: AppConfig = {
    model: derivedModel,
    apiKey: apiKeyForProvider,
    provider: derivedProvider,
    baseURL: derivedBaseURL,
    instructions: loadInstructions(instructionsPath, options),
    approvalMode: storedConfig.approvalMode,
  };

  // -----------------------------------------------------------------------
  // First‑run bootstrap: if the configuration file (and/or its containing
  // directory) didn't exist we create them now so that users end up with a
  // materialised ~/.codex/config.json file on first execution.  This mirrors
  // what `saveConfig()` would do but without requiring callers to remember to
  // invoke it separately.
  //
  // We intentionally perform this *after* we have computed the final
  // `config` object so that we can just persist the resolved defaults.  The
  // write operations are guarded by `existsSync` checks so that subsequent
  // runs that already have a config will remain read‑only here.
  // -----------------------------------------------------------------------

  try {
    if (!existsSync(actualConfigPath)) {
      // Ensure the directory exists first.
      const dir = dirname(actualConfigPath);
      if (!existsSync(dir)) {
        mkdirSync(dir, { recursive: true });
      }

      // Persist a minimal config – we include the `model` key but leave it as
      // an empty string so that `loadConfig()` treats it as "unset" and falls
      // back to whatever DEFAULT_MODEL is current at runtime.  This prevents
      // pinning users to an old default after upgrading Codex.
      const ext = extname(actualConfigPath).toLowerCase();
      if (ext === ".yaml" || ext === ".yml") {
        writeFileSync(actualConfigPath, dumpYaml(EMPTY_STORED_CONFIG), "utf-8");
      } else {
        writeFileSync(actualConfigPath, EMPTY_CONFIG_JSON, "utf-8");
      }
    }
  } catch {
    // Silently ignore any errors – failure to persist the defaults shouldn't
    // block the CLI from starting.  A future explicit `codex config` command
    // or `saveConfig()` call can handle (re‑)writing later.
  }

  // Only include the "memory" key if it was explicitly set by the user. This
  // preserves backward‑compatibility with older config files (and our test
  // fixtures) that don't include a "memory" section.
  if (storedConfig.memory !== undefined) {
    config.memory = storedConfig.memory;
  }

  if (storedConfig.fullAutoErrorMode) {
    config.fullAutoErrorMode = storedConfig.fullAutoErrorMode;
  }

  return config;
};

export const saveConfig = (
  config: AppConfig,
  configPath = CONFIG_FILEPATH,
  instructionsPath = INSTRUCTIONS_FILEPATH,
): void => {
  // If the caller passed the default JSON path *and* a YAML config already
  // exists on disk, save back to that YAML file instead to preserve the
  // user's chosen format.
  let targetPath = configPath;
  if (
    configPath === CONFIG_FILEPATH &&
    !existsSync(configPath) &&
    (existsSync(CONFIG_YAML_FILEPATH) || existsSync(CONFIG_YML_FILEPATH))
  ) {
    targetPath = existsSync(CONFIG_YAML_FILEPATH)
      ? CONFIG_YAML_FILEPATH
      : CONFIG_YML_FILEPATH;
  }

  const dir = dirname(targetPath);
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }

  const ext = extname(targetPath).toLowerCase();
  // Create the config object to save
  const configToSave: StoredConfig = {
    model: config.model,
    approvalMode: config.approvalMode,
  };
  if (ext === ".yaml" || ext === ".yml") {
    writeFileSync(targetPath, dumpYaml(configToSave), "utf-8");
  } else {
    writeFileSync(targetPath, JSON.stringify(configToSave, null, 2), "utf-8");
  }

  writeFileSync(instructionsPath, config.instructions, "utf-8");
};
