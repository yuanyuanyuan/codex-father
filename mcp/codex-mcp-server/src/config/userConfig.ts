import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

export type UserMcpConfig = {
  projectRoot?: string;
  jobSh?: string;
  startSh?: string;
  ignoreMcpStartFailures?: boolean;
  forceSkipBase?: boolean;
};

// Minimal TOML reader for a single section and simple key = value pairs
function parseMinimalTomlSection(content: string, sectionNames: string[]): Record<string, string> {
  const lines = content.split(/\r?\n/);
  let inSection = false;
  const out: Record<string, string> = {};
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('[') && line.endsWith(']')) {
      const sec = line.slice(1, -1).trim();
      inSection = sectionNames.includes(sec);
      continue;
    }
    if (!inSection) continue;
    const eq = line.indexOf('=');
    if (eq <= 0) continue;
    const key = line.slice(0, eq).trim();
    let val = line.slice(eq + 1).trim();
    if (val.startsWith('"') && val.endsWith('"')) {
      val = val.slice(1, -1);
    } else if (val.startsWith("'") && val.endsWith("'")) {
      val = val.slice(1, -1);
    }
    out[key] = val;
  }
  return out;
}

export function readUserMcpConfig(): UserMcpConfig | null {
  const file = path.join(os.homedir(), '.codex', 'config.toml');
  if (!fs.existsSync(file)) return null;
  try {
    const content = fs.readFileSync(file, 'utf8');
    const kv = parseMinimalTomlSection(content, [
      'codex_father',
      'codex-father',
      'codex_father.mcp',
    ]);
    const toBool = (s?: string): boolean | undefined => {
      if (!s) return undefined;
      const v = s.toLowerCase();
      if (v === 'true' || v === '1' || v === 'yes') return true;
      if (v === 'false' || v === '0' || v === 'no') return false;
      return undefined;
    };
    return {
      projectRoot: kv['project_root'] || kv['projectRoot'],
      jobSh: kv['job_sh'] || kv['jobSh'],
      startSh: kv['start_sh'] || kv['startSh'],
      ignoreMcpStartFailures: toBool(kv['ignore_mcp_start_failures']),
      forceSkipBase: toBool(kv['force_skip_base']),
    };
  } catch {
    return null;
  }
}

export function applyUserMcpConfigToEnv(): void {
  const cfg = readUserMcpConfig();
  if (!cfg) return;
  if (cfg.projectRoot && !process.env.CODEX_MCP_PROJECT_ROOT) {
    process.env.CODEX_MCP_PROJECT_ROOT = cfg.projectRoot;
  }
  if (cfg.jobSh && !process.env.CODEX_JOB_SH) {
    process.env.CODEX_JOB_SH = cfg.jobSh;
  }
  if (cfg.startSh && !process.env.CODEX_START_SH) {
    process.env.CODEX_START_SH = cfg.startSh;
  }
  if (cfg.ignoreMcpStartFailures !== undefined && !process.env.CODEX_IGNORE_MCP_START_FAILURES) {
    process.env.CODEX_IGNORE_MCP_START_FAILURES = cfg.ignoreMcpStartFailures ? '1' : '0';
  }
  if (cfg.forceSkipBase !== undefined && !process.env.FORCE_SKIP_BASE) {
    process.env.FORCE_SKIP_BASE = cfg.forceSkipBase ? '1' : '0';
  }
}
