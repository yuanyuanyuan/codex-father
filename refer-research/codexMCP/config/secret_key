w=(ish4zx0vcnywdkz*f$#(=n9hl6fog7$6ou6+nqkj2$207%p