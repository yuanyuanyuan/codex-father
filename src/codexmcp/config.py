"""Configuration management for CodexMCP."""

from __future__ import annotations

import os

from .logging_cfg import logger


class Config:
    """Configuration singleton with environment variable support."""

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._init()
        return cls._instance

    def _init(self):
        """Initialize configuration from environment variables."""
        # Cache settings
        self.cache_enabled = os.environ.get("CODEXMCP_CACHE_ENABLED", "1").lower() in (
            "1",
            "true",
            "yes",
        )
        self.cache_ttl = int(os.environ.get("CODEXMCP_CACHE_TTL", "3600"))

        # Retry settings
        self.max_retries = int(os.environ.get("CODEXMCP_MAX_RETRIES", "3"))
        self.retry_backoff = float(os.environ.get("CODEXMCP_RETRY_BACKOFF", "2.0"))

        # Model settings
        self.default_model = os.environ.get("CODEXMCP_DEFAULT_MODEL", "o4-mini")
        self.default_temperature = float(os.environ.get("CODEXMCP_DEFAULT_TEMP", "0.2"))
        self.default_max_tokens = int(
            os.environ.get("CODEXMCP_DEFAULT_MAX_TOKENS", "4096")
        )

        # Log configuration values
        self._log_config()

    def _log_config(self):
        """Log the current configuration."""
        logger.info("CodexMCP Configuration:")
        logger.info(
            f"  Cache: {'enabled' if self.cache_enabled else 'disabled'} (TTL: {self.cache_ttl}s)"
        )
        logger.info(f"  Retries: max={self.max_retries}, backoff={self.retry_backoff}")
        logger.info(f"  Default model: {self.default_model}")
        logger.info(
            f"  Default params: temp={self.default_temperature}, max_tokens={self.default_max_tokens}"
        )


# Global singleton instance
config = Config()
