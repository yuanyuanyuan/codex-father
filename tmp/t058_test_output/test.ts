import { detectVersion } from "./version-detector.ts";

const inputs = ["v1.2.3", "1.2", "foo", "v0.0.0"];
for (const s of inputs) {
  console.log(s, "=>", detectVersion(s));
}

