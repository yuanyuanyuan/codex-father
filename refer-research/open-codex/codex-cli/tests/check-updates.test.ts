import { describe, it, expect, vi } from "vitest";
import {
  checkForUpdates,
  checkOutdated,
  getNPMCommandPath,
} from "../src/utils/check-updates.js";
import { execFile } from "node:child_process";
import { join } from "node:path";
import { CONFIG_DIR } from "src/utils/config.js";
import { beforeEach } from "node:test";

vi.mock("which", () => ({
  default: vi.fn(() => "/usr/local/bin/npm"),
}));

vi.mock("child_process", () => ({
  execFile: vi.fn((_cmd, _args, _opts, callback) => {
    const stdout = JSON.stringify({
      "open-codex": {
        current: "1.0.0",
        latest: "2.0.0",
      },
    });
    callback?.(null, stdout, "");
    return {} as any;
  }),
}));

let memfs: Record<string, string> = {};

vi.mock("node:fs/promises", async (importOriginal) => ({
  ...(await importOriginal()),
  readFile: async (path: string) => {
    if (memfs[path] === undefined) {
      throw new Error("ENOENT");
    }
    return memfs[path];
  },
}));

beforeEach(() => {
  memfs = {}; // reset inâ€‘memory store
});

describe("Check for updates", () => {
  it("should return the path to npm", async () => {
    const npmPath = await getNPMCommandPath();
    expect(npmPath).toBeDefined();
  });

  it("should return undefined if npm is not found", async () => {
    vi.mocked(await import("which")).default.mockImplementationOnce(() => {
      throw new Error("not found");
    });

    const npmPath = await getNPMCommandPath();
    expect(npmPath).toBeUndefined();
  });

  it("should return the return value when package is outdated", async () => {
    const npmPath = await getNPMCommandPath();

    const info = await checkOutdated(npmPath!);
    expect(info).toStrictEqual({
      currentVersion: "1.0.0",
      latestVersion: "2.0.0",
    });
  });

  it("should return undefined when package is not outdated", async () => {
    const npmPath = await getNPMCommandPath();
    vi.mocked(execFile).mockImplementationOnce(
      (_cmd, _args, _opts, callback) => {
        // Simulate the case where the package is not outdated, returning an empty object
        const stdout = JSON.stringify({});
        callback?.(null, stdout, "");
        return {} as any;
      },
    );

    const info = await checkOutdated(npmPath!);
    expect(info).toBeUndefined();
  });

  it("should outputs the update message when package is outdated", async () => {
    const codexStatePath = join(CONFIG_DIR, "update-check.json");
    // Use a fixed early date far in the past to ensure it's always at least 1 day before now
    memfs[codexStatePath] = JSON.stringify({
      lastUpdateCheck: new Date("2000-01-01T00:00:00Z").toUTCString(),
    });
    await checkForUpdates();
    // Spy on console.log to capture output
    const logSpy = vi.spyOn(console, "log").mockImplementation(() => {});
    await checkForUpdates();
    expect(logSpy).toHaveBeenCalled();
    // The last call should be the boxen message
    const lastCallArg = logSpy.mock.calls.at(-1)?.[0];
    expect(lastCallArg).toMatchSnapshot();
  });

  it("should not output the update message when package is not outdated", async () => {
    const codexStatePath = join(CONFIG_DIR, "update-check.json");
    memfs[codexStatePath] = JSON.stringify({
      lastUpdateCheck: new Date().toUTCString(),
    });
    await checkForUpdates();
    // Spy on console.log to capture output
    const logSpy = vi.spyOn(console, "log").mockImplementation(() => {});
    await checkForUpdates();
    expect(logSpy).not.toHaveBeenCalled();
  });
});
