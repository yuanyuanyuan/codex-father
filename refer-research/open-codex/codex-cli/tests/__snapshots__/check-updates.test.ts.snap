// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check for updates > should outputs the update message when package is outdated 1`] = `
"
   ╭──────────────────────────────────────────────────────────╮
   │                                                          │
   │             Update available! 1.0.0 → 2.0.0.             │
   │   To update, run: npm install -g open-codex to update.   │
   │                                                          │
   ╰──────────────────────────────────────────────────────────╯
"
`;
